catboost/libs/fstr/shap_values.cpp

```diff
@@ -1235,6 +1441,601 @@ TVector<TVector<double>> CalcShapValues(
+        binarizedFeatures->emplace_back();
+        indexes->emplace_back();
+        auto& indexesForBlock = indexes->back();
+        indexesForBlock.resize(documentBlockSize * forest.GetTreeCount());
+        auto& binarizedFeaturesForBlock = binarizedFeatures->back();
+        binarizedFeaturesForBlock = MakeQuantizedFeaturesForEvaluator(model, *featuresBlockIterator, blockParams.FirstId, blockParams.LastId);
+
+        model.GetCurrentEvaluator()->CalcLeafIndexes(
+            binarizedFeaturesForBlock.Get(),
+            0, forest.GetTreeCount(),
+            MakeArrayRef(indexesForBlock.data(), binarizedFeaturesForBlock->GetObjectsCount() * forest.GetTreeCount())
+        );
+    }
+}
+
+static TVector<TVector<TVector<double>>> CalcShapValueWithQuantizedData(
+    const TFullModel& model, 
+    const TVector<TIntrusivePtr<NModelEvaluation::IQuantizedData>>& binarizedFeatures,
+    const TVector<TVector<NModelEvaluation::TCalcerIndexType>>& indexes,
+    const TMaybe<TFixedFeatureParams>& fixedFeatureParams,
+    const size_t documentCount,
+    int logPeriod,
+    TShapPreparedTrees* preparedTrees,
+    NPar::TLocalExecutor* localExecutor
+) {
+    CalcShapValuesByLeaf(
+        model,
+        fixedFeatureParams,
+        logPeriod,
+        preparedTrees->CalcInternalValues,
+        localExecutor,
+        preparedTrees
+    );
+    const TModelTrees& forest = *model.ModelTrees;
+    TVector<TVector<TVector<double>>> shapValues(documentCount);
+    const int featuresCount = preparedTrees->CombinationClassFeatures.size();
+    const size_t documentBlockSize = CB_THREAD_LIMIT;
+    for (ui32 startIdx = 0, blockIdx = 0; startIdx < documentCount; startIdx += documentBlockSize, ++blockIdx) {
+        NPar::TLocalExecutor::TExecRangeParams blockParams(startIdx, startIdx + Min(documentBlockSize, documentCount - startIdx));
+        auto binarizedFeaturesBlock = binarizedFeatures[blockIdx];
+        auto& indexesForBlock = indexes[blockIdx];
+        localExecutor->ExecRange([&](ui32 documentIdx) {
+            const size_t documentIdxInBlock = documentIdx - startIdx;
+            auto docIndexes = MakeArrayRef(indexesForBlock.data() + forest.GetTreeCount() * documentIdxInBlock, forest.GetTreeCount());
+            CalcShapValuesForDocumentMulti(
+                model,
+                *preparedTrees,
+                binarizedFeaturesBlock.Get(),
+                fixedFeatureParams,
+                featuresCount,
+                docIndexes,
+                documentIdxInBlock,
+                &shapValues[documentIdx]
+            );
+        }, blockParams, NPar::TLocalExecutor::WAIT_COMPLETE);
+    }
+    return shapValues;
+}
+
+using TInteractionValuesSubset = THashMap<std::pair<size_t, size_t>, TVector<TVector<double>>>;
+using TInteractionValuesFull = TVector<TVector<TVector<TVector<double>>>>;
+
+template <typename TStorageType>
+TVector<TVector<double>>* GetDocumentsByClasses(TStorageType* rank4storage, size_t idx1, size_t idx2);
```

Comment:
should return a reference:

TVector<TVector<double>>& GetDocumentsByClasses(TStorageType* rank4storage, size_t idx1, size_t idx2);